Modmenu produces a NPE when having an unknown Badge entry

Having a mod that has for example the following entry in the `fabric.mod.json` lets modmenu produce a NPE while rendering badges.
```json
"custom": {
    "modmenu": {
      "badges": [
        "liteloader"
      ]
    }
  }
```
This occours, because the `badge` variable in the second method here is `null`:
https://github.com/TerraformersMC/ModMenu/blob/5d640d66a3e182889f84075d8fded0470534dbf5/src/main/java/com/terraformersmc/modmenu/util/mod/ModBadgeRenderer.java#L31-L40
This happens because the `Set<Badge>` contains `null`. After looking from where this set comes, you get to here:
https://github.com/TerraformersMC/ModMenu/blob/5d640d66a3e182889f84075d8fded0470534dbf5/src/main/java/com/terraformersmc/modmenu/util/mod/fabric/FabricMod.java#L92-L131
As you can see, in Line 113 it sets the field `this.badges`, which is returned by the `getBadges` method. As you can see, this code doesn't directly add `null` as a value. Therefore, the value `null` must come from the field `modMenuData.badges`. This field is set in the constructor of `ModMenuData`:
https://github.com/TerraformersMC/ModMenu/blob/5d640d66a3e182889f84075d8fded0470534dbf5/src/main/java/com/terraformersmc/modmenu/util/mod/fabric/FabricMod.java#L337-L341
This just calls `Badges.convert()`, which takes a `Set<String>` and gives back a `Set<Badge>`:
https://github.com/TerraformersMC/ModMenu/blob/5d640d66a3e182889f84075d8fded0470534dbf5/src/main/java/com/terraformersmc/modmenu/util/mod/Mod.java#L151-L153
And this also shows us where the `null` comes from: `KEY_MAP::get` can return `null` if there's no key-value pair. And that's percisely what happens in my case.

Note that the usage of `liteloader` here is just because it happened in a modmenu port to [Ornithe](https://github.com/OrnitheMC), a modding toolchain for old minecraft versions. The relevant code path was not changed in the code. Therefore it could also happen in modern modmenu.
(I also fist asked there, <https://github.com/OrnitheMC/modmenu/issues/4>, and we determined that upstream also has the issue, and therefore I think that it should be fixed in upstream)

A crashlog from Ornithe:
```txt
Caused by: java.lang.NullPointerException
	at com.terraformersmc.modmenu.util.mod.ModBadgeRenderer.drawBadge(ModBadgeRenderer.java:33) ~[modmenu-0.2.0+mc1.12.2.jar:?]
	at com.terraformersmc.modmenu.util.mod.ModBadgeRenderer.lambda$draw$0(ModBadgeRenderer.java:29) ~[modmenu-0.2.0+mc1.12.2.jar:?]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:1.8.0_392]
	at com.terraformersmc.modmenu.util.mod.ModBadgeRenderer.draw(ModBadgeRenderer.java:29) ~[modmenu-0.2.0+mc1.12.2.jar:?]
	at com.terraformersmc.modmenu.gui.widget.entries.ModListEntry.render(ModListEntry.java:75) ~[modmenu-0.2.0+mc1.12.2.jar:?]
	at net.minecraft.client.gui.widget.EntryListWidget.renderEntry(EntryListWidget.java:28) ~[1.12.2-TRv0.8.2-named-711666161528F3329D7A8C8C611232C03E2D58B845ABA1B8F40897F5800EAE50.jar:?]
	at com.terraformersmc.modmenu.gui.widget.ModListWidget.renderList(ModListWidget.java:246) ~[modmenu-0.2.0+mc1.12.2.jar:?]
	at net.minecraft.client.gui.widget.ListWidget.render(ListWidget.java:194) ~[1.12.2-TRv0.8.2-named-711666161528F3329D7A8C8C611232C03E2D58B845ABA1B8F40897F5800EAE50.jar:?]
	at com.terraformersmc.modmenu.gui.ModsScreen.render(ModsScreen.java:321) ~[modmenu-0.2.0+mc1.12.2.jar:?]
	at net.minecraft.client.render.GameRenderer.render(GameRenderer.java:947) ~[1.12.2-TRv0.8.2-named-711666161528F3329D7A8C8C611232C03E2D58B845ABA1B8F40897F5800EAE50.jar:?]
	... 6 more
```

This issue also comes with further questions:
- Should there be some kind of API to register badge? This would, at least partially, remove the issue, as then one mod counld register the badge `liteloader` and define it's colors and text (see what skyrising used for an example here: <https://github.com/skyrising/ModMenu/commit/385c15f8197f83f0603ec23db98a1a6b00bd7a64>)
- This would however still not 100% fix the issue, `Badges.convert()` should at least either throw an exception if there's any `null` in the set (that could happen if a mod forgot to register a badge), or silently filter out `null`s in the set. This could be done by filtering the stream: `.filter(Objects::nonNull)`.

--- a/src/main/java/com/terraformersmc/modmenu/util/mod/Mod.java
+++ b/src/main/java/com/terraformersmc/modmenu/util/mod/Mod.java
@@ -1,5 +1,7 @@
 package com.terraformersmc.modmenu.util.mod;
 
+import com.google.common.base.Predicates;
+import com.terraformersmc.modmenu.ModMenu;
 import com.terraformersmc.modmenu.api.UpdateChecker;
 import com.terraformersmc.modmenu.api.UpdateInfo;
 import com.terraformersmc.modmenu.config.ModMenuConfig;
@@ -166,8 +168,17 @@ public interface Mod {
 			return this.fillColor;
 		}
 
-		public static Set<Badge> convert(Set<String> badgeKeys) {
-			return badgeKeys.stream().map(KEY_MAP::get).collect(Collectors.toSet());
+		public static Set<Badge> convert(Set<String> badgeKeys, String modId) {
+			return badgeKeys.stream()
+					.map(key -> {
+						if (!KEY_MAP.containsKey(key)) {
+							ModMenu.LOGGER.warn("Skipping unknown badge key '{}' specified by mod '{}'", key, modId);
+						}
+
+						return KEY_MAP.get(key);
+					})
+					.filter(Objects::nonNull)
+					.collect(Collectors.toSet());
 		}
 
 		static {

--- a/src/main/java/com/terraformersmc/modmenu/util/mod/fabric/FabricMod.java
+++ b/src/main/java/com/terraformersmc/modmenu/util/mod/fabric/FabricMod.java
@@ -51,7 +51,9 @@ public class FabricMod implements Mod {
 		this.container = modContainer;
 		this.metadata = modContainer.getMetadata();
 
-		if ("minecraft".equals(metadata.getId()) || "java".equals(metadata.getId())) {
+		String id = metadata.getId();
+
+		if ("minecraft".equals(id) || "java".equals(id)) {
 			allowsUpdateChecks = false;
 		}
 
@@ -77,13 +79,13 @@ public class FabricMod implements Mod {
 								CustomValueUtil.getString("icon", parentObj),
 								CustomValueUtil.getStringSet("badges", parentObj).orElse(new HashSet<>())
 						);
-						if (parentId.orElse("").equals(this.metadata.getId())) {
+						if (parentId.orElse("").equals(id)) {
 							parentId = Optional.empty();
 							parentData = null;
 							throw new RuntimeException("Mod declared itself as its own parent");
 						}
 					} catch (Throwable t) {
-						LOGGER.error("Error loading parent data from mod: " + metadata.getId(), t);
+						LOGGER.error("Error loading parent data from mod: " + id, t);
 					}
 				}
 			}
@@ -94,11 +96,11 @@ public class FabricMod implements Mod {
 		this.modMenuData = new ModMenuData(
 				badgeNames,
 				parentId,
-				parentData
+				parentData,
+				id
 		);
 
 		/* Hardcode parents and badges for Fabric API & Fabric Loader */
-		String id = metadata.getId();
 		if (id.startsWith("fabric") && metadata.containsCustomValue("fabric-api:module-lifecycle")) {
 			if (FabricLoader.getInstance().isModLoaded("fabric-api") || !FabricLoader.getInstance().isModLoaded("fabric")) {
 				modMenuData.fillParentIfEmpty("fabric-api");
@@ -365,8 +367,8 @@ public class FabricMod implements Mod {
 		private @Nullable
 		final DummyParentData dummyParentData;
 
-		public ModMenuData(Set<String> badges, Optional<String> parent, DummyParentData dummyParentData) {
-			this.badges = Badge.convert(badges);
+		public ModMenuData(Set<String> badges, Optional<String> parent, DummyParentData dummyParentData, String id) {
+			this.badges = Badge.convert(badges, id);
 			this.parent = parent;
 			this.dummyParentData = dummyParentData;
 		}
@@ -413,7 +415,7 @@ public class FabricMod implements Mod {
 				this.name = name;
 				this.description = description;
 				this.icon = icon;
-				this.badges = Badge.convert(badges);
+				this.badges = Badge.convert(badges, id);
 			}
 
 			public String getId() {

